storage/metamorphic: TestPebbleEquivalence failed

storage/metamorphic.TestPebbleEquivalence [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3688252&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3688252&tab=artifacts#/) on master @ [ed35bbe8f6e5947db6e3f2423d1aa9dde1d15804](https://github.com/cockroachdb/cockroach/commits/ed35bbe8f6e5947db6e3f2423d1aa9dde1d15804):


```
=== RUN   TestPebbleEquivalence
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPebbleEquivalence503536829
    test_log_scope.go:80: use -show-logs to present logs inline
first run output file: /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence879817464/output.meta
=== CONT  TestPebbleEquivalence
    meta_test.go:205: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPebbleEquivalence503536829
--- FAIL: TestPebbleEquivalence (525.51s)
=== RUN   TestPebbleEquivalence/engine/standard=7
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence879817464/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=7214466125/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=7112879432/store
    generator.go:415: mismatching output at line 42290, operation index 42190: expected truncated range to delete = "hhkmxpkvenw" - "xjtrclkeqjmndk", deleted keys = "hkuvyjjaatd", "hqxhppwu", "iusewhth", "ivxigflfcqfnbrp", "jehggympkat", "jguxpsbhh", "jjgqpnccsb", "jkgltwsfiro", "wfchcsqfgmv", "wpheiqgnudi", "wsmrtqsrbflwdwu", "wvqhpvxbfhji", "wyjxdcxcjxvova", "xdjfkdinhacx", "xhmkoowsauf"
        , got error: WriteTooOldError: write at timestamp 0.000003821,0 too old; wrote at 0.000003821,1
    --- FAIL: TestPebbleEquivalence/engine/standard=7 (30.93s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPebbleEquivalence.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

storage/metamorphic.TestPebbleEquivalence [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3796973&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3796973&tab=artifacts#/) on master @ [38fd4064f40c574d0b2aee02f414787a2760fcec](https://github.com/cockroachdb/cockroach/commits/38fd4064f40c574d0b2aee02f414787a2760fcec):


```
    --- FAIL: TestPebbleEquivalence/engine/standard=3 (0.56s)
=== RUN   TestPebbleEquivalence/engine/standard=6
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=63612728184/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=64076981756/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=6 (0.54s)
=== RUN   TestPebbleEquivalence/engine/standard=11
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=112033679695/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=112623076582/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=11 (0.51s)
=== RUN   TestPebbleEquivalence/engine/standard=13
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=133651881231/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=133667999216/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=13 (0.52s)
=== RUN   TestPebbleEquivalence/engine/standard=14
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=142099917635/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=142597173835/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=14 (0.61s)
=== RUN   TestPebbleEquivalence/engine/standard=1
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=1564302503/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=13312171447/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=1 (0.53s)
=== RUN   TestPebbleEquivalence/engine/standard=4
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=41578846370/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=42618838742/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=4 (0.52s)
=== RUN   TestPebbleEquivalence/engine/standard=10
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence969131495/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=102895803966/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=102007179802/store
    generator.go:415: mismatching output at line 1051, operation index 951: expected ok, span = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=10 (0.54s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPebbleEquivalence.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

storage/metamorphic.TestPebbleEquivalence [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4007671&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4007671&tab=artifacts#/) on master @ [01c0c739dd83b3e662d715404f2193d747f4fb9e](https://github.com/cockroachdb/cockroach/commits/01c0c739dd83b3e662d715404f2193d747f4fb9e):


```
=== RUN   TestPebbleEquivalence
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPebbleEquivalence614807209
    test_log_scope.go:80: use -show-logs to present logs inline
first run output file: /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence4152504816/output.meta
=== CONT  TestPebbleEquivalence
    meta_test.go:205: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPebbleEquivalence614807209
--- FAIL: TestPebbleEquivalence (569.62s)
=== RUN   TestPebbleEquivalence/engine/standard=7
check file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence4152504816/output.meta
output file = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=71494617560/output.meta
store path = /go/src/github.com/cockroachdb/cockroach/artifacts/TestPebbleEquivalence_engine_standard=71499742639/store
    generator.go:415: mismatching output at line 34699, operation index 34599: expected truncated range to delete = "ansbelemqdiao" - "rpnafxcq", deleted keys = "aygfkgtvew", "bfikvmhno", "bjulaqilwg", "brmaqmpnckcyts", "bstkksnpfusjg", "buccjuwesmxgd", "burmwfklkoqe", "bvqyhmtc", "bysfouaphiwh", "igpenuxowhwqtb", "jdtufxsry", "jhorvifrqcyg", "jldxpdhecnqiw", "jqsplqmrt", "jursvdaufqg", "kfnpfbpxxbv", "kykypcdtxsfblg", "llhstwgyvwka", "lvtwvdgikee", "mpjfhhvgoqu", "mwiqjleyv", "nfarsghatmc", "nmvsotbty", "nxfteiebmuk", "oadvnsupy", "oelfojdh", "ogdwxerwx", "osssqtviaece", "qbojjvjrlvp", "qhtrlctoahm", "qtnwdbxmcpsc"
        , got error: WriteTooOldError: write at timestamp 0.000003135,0 too old; wrote at 0.000003135,1
    --- FAIL: TestPebbleEquivalence/engine/standard=7 (23.02s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPebbleEquivalence.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/storage/metamorphic/operands.go
+++ b/pkg/storage/metamorphic/operands.go
@@ -345,6 +345,31 @@ func (t *txnGenerator) forEachConflict(
 	}
 }
 
+// truncateSpanForConflicts truncates [key, endKey) to a sub-span that does
+// not conflict with any in-flight writes. If no such span is found, an empty
+// span (i.e. key >= endKey) is returned. Callers are expected to handle that
+// case gracefully.
+func (t *txnGenerator) truncateSpanForConflicts(
+	w readWriterID, txn txnID, key, endKey roachpb.Key,
+) roachpb.Span {
+	// forEachConflict is guaranteed to iterate over conflicts in key order,
+	// with the lowest conflicting key first. Find the first conflict and
+	// truncate the span to that range.
+	t.forEachConflict(w, txn, key, endKey, func(conflict roachpb.Span) bool {
+		if conflict.ContainsKey(key) {
+			key = append([]byte(nil), conflict.EndKey...)
+			return true
+		}
+		endKey = conflict.Key
+		return false
+	})
+	result := roachpb.Span{
+		Key:    key,
+		EndKey: endKey,
+	}
+	return result
+}
+
 func (t *txnGenerator) addWrittenKeySpan(
 	w readWriterID, txn txnID, key roachpb.Key, endKey roachpb.Key,
 ) {
@@ -421,6 +446,10 @@ func (t *txnGenerator) addWrittenKeySpan(
 }
 
 func (t *txnGenerator) trackTransactionalWrite(w readWriterID, txn txnID, key, endKey roachpb.Key) {
+	if len(endKey) > 0 && key.Compare(endKey) >= 0 {
+		// No-op.
+		return
+	}
 	t.addWrittenKeySpan(w, txn, key, endKey)
 	if w == "engine" {
 		return

--- a/pkg/storage/metamorphic/operations.go
+++ b/pkg/storage/metamorphic/operations.go
@@ -271,6 +271,11 @@ type mvccDeleteRangeOp struct {
 func (m mvccDeleteRangeOp) run(ctx context.Context) string {
 	txn := m.m.getTxn(m.txn)
 	writer := m.m.getReadWriter(m.writer)
+	if m.key.Compare(m.endKey) >= 0 {
+		// Empty range. No-op.
+		return "no-op due to no non-conflicting key range"
+	}
+
 	txn.Sequence++
 
 	keys, _, _, err := storage.MVCCDeleteRange(ctx, writer, nil, m.key, m.endKey, 0, txn.WriteTimestamp, txn, true)
@@ -295,7 +300,6 @@ func (m mvccDeleteRangeOp) run(ctx context.Context) string {
 
 type mvccClearTimeRangeOp struct {
 	m         *metaTestRunner
-	writer    readWriterID
 	key       roachpb.Key
 	endKey    roachpb.Key
 	startTime hlc.Timestamp
@@ -303,14 +307,16 @@ type mvccClearTimeRangeOp struct {
 }
 
 func (m mvccClearTimeRangeOp) run(ctx context.Context) string {
-	writer := m.m.getReadWriter(m.writer)
-	useTBI := m.writer == "engine"
-	span, err := storage.MVCCClearTimeRange(ctx, writer, &enginepb.MVCCStats{}, m.key, m.endKey,
-		m.startTime, m.endTime, math.MaxInt64, math.MaxInt64, useTBI)
+	if m.key.Compare(m.endKey) >= 0 {
+		// Empty range. No-op.
+		return "no-op due to no non-conflicting key range"
+	}
+	span, err := storage.MVCCClearTimeRange(ctx, m.m.engine, &enginepb.MVCCStats{}, m.key, m.endKey,
+		m.startTime, m.endTime, math.MaxInt64, math.MaxInt64, true /* useTBI */)
 	if err != nil {
 		return fmt.Sprintf("error: %s", err)
 	}
-	return fmt.Sprintf("ok, span = %v", span)
+	return fmt.Sprintf("ok, deleted span = %s - %s, resumeSpan = %v", m.key, m.endKey, span)
 }
 
 type mvccDeleteOp struct {
@@ -678,11 +684,15 @@ type clearRangeOp struct {
 func (c clearRangeOp) run(ctx context.Context) string {
 	// ClearRange calls in Cockroach usually happen with boundaries demarcated
 	// using unversioned keys, so mimic the same behavior here.
+	if c.key.Compare(c.endKey) >= 0 {
+		// Empty range. No-op.
+		return "no-op due to no non-conflicting key range"
+	}
 	err := c.m.engine.ClearMVCCRangeAndIntents(c.key, c.endKey)
 	if err != nil {
 		return fmt.Sprintf("error: %s", err.Error())
 	}
-	return "ok"
+	return fmt.Sprintf("deleted range = %s - %s", c.key, c.endKey)
 }
 
 type compactOp struct {
@@ -887,11 +897,9 @@ var opGenerators = []opGenerator{
 			if endKey.Compare(key) < 0 {
 				key, endKey = endKey, key
 			}
-			// forEachConflict is guaranteed to iterate
-			m.txnGenerator.forEachConflict(writer, txn, key, endKey, func(conflict roachpb.Span) bool {
-				endKey = conflict.Key
-				return false
-			})
+			truncatedSpan := m.txnGenerator.truncateSpanForConflicts(writer, txn, key, endKey)
+			key = truncatedSpan.Key
+			endKey = truncatedSpan.EndKey
 
 			// Track this write in the txn generator. This ensures the batch will be
 			// committed before the transaction is committed
@@ -918,15 +926,17 @@ var opGenerators = []opGenerator{
 	{
 		name: "mvcc_clear_time_range",
 		generate: func(ctx context.Context, m *metaTestRunner, args ...string) mvccOp {
-			writer := readWriterID(args[0])
-			key := m.keyGenerator.parse(args[1]).Key
-			endKey := m.keyGenerator.parse(args[2]).Key
-			startTime := m.pastTSGenerator.parse(args[3])
+			key := m.keyGenerator.parse(args[0]).Key
+			endKey := m.keyGenerator.parse(args[1]).Key
+			startTime := m.pastTSGenerator.parse(args[2])
 			endTime := m.pastTSGenerator.parse(args[3])
 
 			if endKey.Compare(key) < 0 {
 				key, endKey = endKey, key
 			}
+			truncatedSpan := m.txnGenerator.truncateSpanForConflicts("engine", "", key, endKey)
+			key = truncatedSpan.Key
+			endKey = truncatedSpan.EndKey
 			if endTime.Less(startTime) {
 				startTime, endTime = endTime, startTime
 			}
@@ -937,18 +947,13 @@ var opGenerators = []opGenerator{
 			endTime = endTime.Next()
 			return &mvccClearTimeRangeOp{
 				m:         m,
-				writer:    writer,
 				key:       key,
 				endKey:    endKey,
 				startTime: startTime,
 				endTime:   endTime,
 			}
 		},
-		dependentOps: func(m *metaTestRunner, args ...string) []opReference {
-			return closeItersOnBatch(m, readWriterID(args[0]))
-		},
 		operands: []operandType{
-			operandReadWriter,
 			operandMVCCKey,
 			operandMVCCKey,
 			operandPastTS,
@@ -1336,6 +1341,10 @@ var opGenerators = []opGenerator{
 				// standardize behavior.
 				endKey = endKey.Next()
 			}
+
+			truncatedSpan := m.txnGenerator.truncateSpanForConflicts("engine", "", key, endKey)
+			key = truncatedSpan.Key
+			endKey = truncatedSpan.EndKey
 
 			return &clearRangeOp{
 				m:      m,

